core/state-db/src/noncanonical.rs

```diff
@@ -164,85 +171,117 @@ impl<BlockHash: Hash, Key: Hash> NonCanonicalOverlay<BlockHash, Key> {
 		trace!(target: "state-db", "Inserted uncanonicalized changeset {}.{} ({} inserted, {} deleted)", number, index, journal_record.inserted.len(), journal_record.deleted.len());
 		let journal_record = journal_record.encode();
 		commit.meta.inserted.push((journal_key, journal_record));
+		self.pending_insertions.push(hash.clone());
 		Ok(commit)
 	}
 
-	fn discard(
-		levels: &mut [Vec<BlockOverlay<BlockHash, Key>>],
+	fn discard_descendants(
+		levels: &mut VecDeque<Vec<BlockOverlay<BlockHash, Key>>>,
+		index: usize,
 		parents: &mut HashMap<BlockHash, BlockHash>,
-		discarded_journals: &mut Vec<Vec<u8>>,
-		number: u64,
 		hash: &BlockHash,
 	) {
-		if let Some((level, sublevels)) = levels.split_first_mut() {
+		let mut discarded = Vec::new();
+		if let Some(level) = levels.get_mut(index) {
 			level.retain(|ref overlay| {
 				let parent = parents.get(&overlay.hash).expect("there is a parent entry for each entry in levels; qed").clone();
 				if parent == *hash {
 					parents.remove(&overlay.hash);
-					discarded_journals.push(overlay.journal_key.clone());
-					Self::discard(sublevels, parents, discarded_journals, number + 1, &overlay.hash);
+					discarded.push(overlay.hash.clone());
 					false
 				} else {
 					true
 				}
 			});
 		}
+		for hash in discarded.into_iter() {
+			Self::discard_descendants(levels, index + 1, parents, &hash);
+		}
+	}
+
+	fn discard_journals(&self, level_index: usize, discarded_journals: &mut Vec<Vec<u8>>, hash: &BlockHash) {
```

Comment:
is `discarded_journals` actually mutated in this function?

Reply:
it is also not read from as far as I see

Reply:
good catch